File: 03 March/03-06/trailTraversal.js
function navigateTrail(map) {
  const start = 'C';
  const traversableOrGoal = ['T', 'G'];
  const currentLocation = { };
  let trailMoves = "";

  for (const verticalCoordinate in map) {
    const horizontalCoordinate = map[verticalCoordinate].indexOf(start);
    
    if (horizontalCoordinate !== -1) {
      currentLocation['type'] = start;
      currentLocation['vertical'] = Number(verticalCoordinate);
      currentLocation['horizontal'] = horizontalCoordinate;
    }
  }

/* Solution using four if statements, one for each direction
  while (currentLocation['type'] !== 'G') {
    const indexOfOneRowUp = currentLocation['vertical'] - 1;
    const upLocation = indexOfOneRowUp >= 0 ? map[indexOfOneRowUp][currentLocation['horizontal']] : undefined;
    if (trailMoves[trailMoves.length - 1] !== "D" && traversableOrGoal.includes(upLocation)) {
      currentLocation['type'] = upLocation;
      currentLocation['vertical'] -= 1;
      trailMoves += "U";
      continue;
    }

    const indexOfOneRowDown = currentLocation['vertical'] + 1;
    const downLocation = indexOfOneRowDown < map.length ? map[indexOfOneRowDown][currentLocation['horizontal']] : undefined;
    if (trailMoves[trailMoves.length - 1] !== "U" && traversableOrGoal.includes(downLocation)) {
      currentLocation['type'] = downLocation;
      currentLocation['vertical'] += 1;
      trailMoves += "D";
      continue;
    }    

    const indexOfOneColumnLeft = currentLocation['horizontal'] - 1;
    const leftLocation = indexOfOneColumnLeft >= 0 ? map[currentLocation['vertical']][indexOfOneColumnLeft] : undefined;
    if (trailMoves[trailMoves.length - 1] !== "R" && traversableOrGoal.includes(leftLocation)) {
      currentLocation['type'] = leftLocation;
      currentLocation['horizontal'] -= 1;
      trailMoves += "L";
      continue;
    }

    const indexOfOneColumnRight = currentLocation['horizontal'] + 1;
    const rightLocation = indexOfOneColumnRight < map[currentLocation['vertical']].length ? map[currentLocation['vertical']][indexOfOneColumnRight] : undefined;
    if (trailMoves[trailMoves.length - 1] !== "L" && traversableOrGoal.includes(rightLocation)) {
      currentLocation['type'] = rightLocation;
      currentLocation['horizontal'] += 1;
      trailMoves += "R";
    }
  }
*/

// A more elegant condensed solution using an object for directions
  const directions = {
      'U': {verticalOffset: -1, horizontalOffset: 0, oppositeDirection: "D"},
      'D': {verticalOffset: 1, horizontalOffset: 0, oppositeDirection: "U"},
      'L': {verticalOffset: 0, horizontalOffset: -1, oppositeDirection: "R"},
      'R': {verticalOffset: 0, horizontalOffset: 1, oppositeDirection: "L"}
  };

  while (currentLocation['type'] !== 'G') {
    for (const [direction, {verticalOffset, horizontalOffset, oppositeDirection}] of Object.entries(directions)) {
      const targetVertical = currentLocation['vertical'] + verticalOffset;
      const targetHorizontal = currentLocation['horizontal'] + horizontalOffset;
      const targetLocation = map[targetVertical]?.[targetHorizontal];

      if (traversableOrGoal.includes(targetLocation) && !trailMoves.endsWith(oppositeDirection)) {
        currentLocation['type'] = targetLocation;
        currentLocation['vertical'] = targetVertical;
        currentLocation['horizontal'] = targetHorizontal;
        trailMoves += direction;
        break;
      }
    }
  }

  return trailMoves;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. navigateTrail(["-CT--", "--T--", "--TT-", "---T-", "---G-"]) should return "RDDRDD".
2. navigateTrail(["-----", "--TTG", "--T--", "--T--", "CTT--"]) should return "RRUUURR".
3. navigateTrail(["-C----", "TT----", "T-----", "TTTTT-", "----G-"]) should return "DLDDRRRRD".
4. navigateTrail(["--------", "-CTTT---", "----T---", "---GT---", "--------"]) should return "RRRDDL".
5. navigateTrail(["TTTTTTT-", "T-----T-", "T-----T-", "TTTT--TG", "---C----"]) should return "ULLLUUURRRRRRDDDR".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);

File: 03 March/03-06/trail_traversal.py
def navigate_trail(map):
    start = 'C'
    traversable_or_goal = ['T', 'G']
    current_location = { }
    trail_moves = ""

    for vertical_coordinate, row in enumerate(map):
        try:
            horizontal_coordinate = row.index(start)
        except ValueError:
            pass
        else:
            current_location['type'] = start
            current_location['vertical'] = vertical_coordinate
            current_location['horizontal'] = horizontal_coordinate

# Solution using four if statements, one for each direction
#    while(current_location['type'] != 'G'):
#        index_of_one_row_up = current_location['vertical'] - 1
#        up_location = map[index_of_one_row_up][current_location['horizontal']] if index_of_one_row_up >= 0 else None
#        if not trail_moves.endswith("D") and up_location in traversable_or_goal:
#            current_location['type'] = up_location
#            current_location['vertical'] -= 1
#            trail_moves += "U"
#            continue
#
#        index_of_one_row_down = current_location['vertical'] + 1
#        down_location = map[index_of_one_row_down][current_location['horizontal']] if index_of_one_row_down < len(map) else None
#        if not trail_moves.endswith("U") and down_location in traversable_or_goal:
#            current_location['type'] = down_location
#            current_location['vertical'] += 1
#            trail_moves += "D"
#            continue
#
#        index_of_one_column_left = current_location['horizontal'] - 1
#        left_location = map[current_location['vertical']][index_of_one_column_left] if index_of_one_column_left >= 0 else None
#        if not trail_moves.endswith("R") and left_location in traversable_or_goal:
#            current_location['type'] = left_location
#            current_location['horizontal'] -= 1
#            trail_moves += "L"
#            continue
#
#        index_of_one_column_right = current_location['horizontal'] + 1
#        right_location = map[current_location['vertical']][index_of_one_column_right] if index_of_one_column_right < len(map[current_location['vertical']]) else None
#        if not trail_moves.endswith("L") and right_location in traversable_or_goal:
#            current_location['type'] = right_location
#            current_location['horizontal'] += 1
#            trail_moves += "R"

# A more elegant condensed solution using a dictionary for directions
    directions = {
        'U': (-1, 0, "D"),
        'D': (1, 0, "U"),
        'L': (0, -1, "R"),
        'R': (0, 1, "L")
    }

    while(current_location['type'] != 'G'):
        for direction, (vertical_offset, horizontal_offset, opposite_direction) in directions.items():
            target_vertical = current_location['vertical'] + vertical_offset
            target_horizontal = current_location['horizontal'] + horizontal_offset

            if 0 <= target_vertical < len(map) \
            and 0 <= target_horizontal < len(map[current_location['vertical']]) \
            and map[target_vertical][target_horizontal] in traversable_or_goal \
            and not trail_moves.endswith(opposite_direction):
                current_location['type'] = map[target_vertical][target_horizontal]
                current_location['vertical'] = target_vertical
                current_location['horizontal'] = target_horizontal
                trail_moves += direction
                break

    return trail_moves

# --- TEST SUITE ---

tests_text = r'''
1. navigate_trail(["-CT--", "--T--", "--TT-", "---T-", "---G-"]) should return "RDDRDD".
2. navigate_trail(["-----", "--TTG", "--T--", "--T--", "CTT--"]) should return "RRUUURR".
3. navigate_trail(["-C----", "TT----", "T-----", "TTTTT-", "----G-"]) should return "DLDDRRRRD".
4. navigate_trail(["--------", "-CTTT---", "----T---", "---GT---", "--------"]) should return "RRRDDL".
5. navigate_trail(["TTTTTTT-", "T-----T-", "T-----T-", "TTTT--TG", "---C----"]) should return "ULLLUUURRRRRRDDDR".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)